File: feedback-service/src/main/java/com/gl/learnloop/feedback/service/FeedbackService.java
package com.gl.learnloop.feedback.service;

import com.gl.learnloop.feedback.client.MatchFeignClient;
import com.gl.learnloop.feedback.client.UserFeignClient;
import com.gl.learnloop.feedback.dto.FeedbackRequest;
import com.gl.learnloop.feedback.entity.Feedback;
import com.gl.learnloop.feedback.repository.FeedbackRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class FeedbackService {

    private final FeedbackRepository feedbackRepository;
    private final UserFeignClient userFeignClient;
    private final MatchFeignClient matchFeignClient;

    public FeedbackService(
            FeedbackRepository feedbackRepository,
            UserFeignClient userFeignClient,
            MatchFeignClient matchFeignClient) {

        this.feedbackRepository = feedbackRepository;
        this.userFeignClient = userFeignClient;
        this.matchFeignClient = matchFeignClient;
    }

    // Create Feedback
    public Feedback createFeedback(FeedbackRequest request) {

        // Check if user exists
        try {
            userFeignClient.getUserById(request.getUserId());
        } catch (Exception e) {
            throw new RuntimeException("User not found");
        }

        // Check if match exists
        try {
            matchFeignClient.getMatchById(request.getMatchId());
        } catch (Exception e) {
            throw new RuntimeException("Match not found");
        }

        Feedback feedback = new Feedback();

        feedback.setUserId(request.getUserId());
        feedback.setMatchId(request.getMatchId());
        feedback.setRating(request.getRating());
        feedback.setComment(request.getComment());

        return feedbackRepository.save(feedback);
    }

    // Get all Feedback
    public List<Feedback> getAllFeedback() {

        return feedbackRepository.findAll();
    }

    // Get Feedback by ID
    public Feedback getFeedbackById(Long id) {

        return feedbackRepository.findById(id)
                .orElseThrow(() ->
                        new RuntimeException("Feedback not found"));
    }

    // Get Feedback by User ID
    public List<Feedback> getFeedbackByUserId(Long userId) {

        return feedbackRepository.findByUserId(userId);
    }

    // Get Feedback by Match ID
    public List<Feedback> getFeedbackByMatchId(Long matchId) {

        return feedbackRepository.findByMatchId(matchId);
    }

    // Update Feedback
    public Feedback updateFeedback(
            Long id,
            FeedbackRequest request) {

        Feedback feedback = feedbackRepository.findById(id)
                .orElseThrow(() ->
                        new RuntimeException("Feedback not found"));

        feedback.setRating(request.getRating());
        feedback.setComment(request.getComment());

        return feedbackRepository.save(feedback);
    }

    // Delete Feedback
    public void deleteFeedback(Long id) {

        if (!feedbackRepository.existsById(id)) {
            throw new RuntimeException("Feedback not found");
        }

        feedbackRepository.deleteById(id);
    }
}

File: frontend/src/services/feedbackService.ts
import api from "./api";
import type {
    Feedback,
    CreateFeedbackRequest,
    UpdateFeedbackRequest,
} from "../types/feedback";

const BASE_PATH = "/api/feedback";

export async function createFeedback(data: CreateFeedbackRequest): Promise<Feedback> {
    const response = await api.post<Feedback>(BASE_PATH, data);
    return response.data;
}

export async function getAllFeedback(): Promise<Feedback[]> {
    const response = await api.get<Feedback[]>(BASE_PATH);
    return response.data;
}

export async function getFeedbackById(id: number): Promise<Feedback> {
    const response = await api.get<Feedback>(`${BASE_PATH}/${id}`);
    return response.data;
}

export async function getFeedbackByUser(userId: number): Promise<Feedback[]> {
    const response = await api.get<Feedback[]>(`${BASE_PATH}/user/${userId}`);
    return response.data;
}

export async function getFeedbackByMatch(matchId: number): Promise<Feedback[]> {
    const response = await api.get<Feedback[]>(`${BASE_PATH}/match/${matchId}`);
    return response.data;
}

export async function updateFeedback(
    id: number,
    data: UpdateFeedbackRequest
): Promise<Feedback> {
    const response = await api.put<Feedback>(`${BASE_PATH}/${id}`, data);
    return response.data;
}

export async function deleteFeedback(id: number): Promise<string> {
    const response = await api.delete<string>(`${BASE_PATH}/${id}`);
    return response.data;
}